libbpfgo/helpers/rwArray.go

```diff
@@ -0,0 +1,111 @@
+package helpers
+
+import (
+	"sync"
+)
+
+// Like RWMutex, RWArray allows for multiple concurrent readers but
+// only a single writer. The writers lock a mutex while the readers
+// are lock free.
+//
+// It is implemented as a sparse array with a fixed capacity. This
+// allows the memory to avoid reallocation.
+//
+// The layout is as follows:
+//
+// +-----------------------+
+// | * | x | * | X | * | x |
+// +-|-------|-------------+
+//   |       |       |
+//   \/      \/      \/
+// Chunk   Chunk    Chunk
+//
+// Chunk:
+// +---------------------+
+// | Val | x | Val | Val |
+// +---------------------+
+
+// Sparse array implementation
+const (
+	// Total capacity of the array will be 1024*1024
+	slotsPerChunk = 1024
+	numChunks = 1024
+)
+
+type RWArray struct {
+	chunks []chunk
+	mux     sync.Mutex
+}
+
+func NewRWArray() RWArray{
+	return RWArray{
+		chunks: make([]chunk, numChunks),
+	}
+}
+
+func (a *RWArray) Put(v interface{}) int {
+	a.mux.Lock()
+	defer a.mux.Unlock()
+
+	for i := 0; i < numChunks; i++ {
+		if j := a.chunks[i].put(v); j != -1 {
+			return i*slotsPerChunk + j
+		}
+	}
+
+	return -1
+}
+
+func (a *RWArray) Remove(index uint) {
+	a.mux.Lock()
+	defer a.mux.Unlock()
+
+	a.chunks[index / slotsPerChunk].free(index % slotsPerChunk)
+}
+
+func (a *RWArray) Get(index uint) interface{} {
+	chunkIdx := index / slotsPerChunk
+	slotIdx := index % slotsPerChunk
+
+	return a.chunks[chunkIdx].slots[slotIdx].value
+}
+
+func (a *RWArray) Capacity() uint {
+	return slotsPerChunk * numChunks
+}
+
+type slot struct {
+	value interface{}
+	used bool
+}
+
+type chunk struct {
+	slots []slot
+	used uint
+}
+
+func (c *chunk) put(v interface{}) int {
+	if c.used == slotsPerChunk {
+		return -1
+	} else if c.used == 0 && c.slots == nil {
```

Comment:
I need to comment more as this is subtle. Since the `Get` path accesses `c.slots` without holding a lock, I need to make sure that `c.used == 0` before touching `c.slots`. I'm relying on Go's short-circuiting behavior here. `c.used == 0` guarantees that no readers exist.